### ISSUE TITLE
Feature request: SuggestWidget mode for less-aggressive completion

### ISSUE THREAD

**Issue by xirzec** [microsoft/monaco-editor#709]
We currently use the monaco-editor inside the Edge DevTools console for command input. One issue we currently face is that the SuggestWidget will always complete with the top item when the user hits enter, even if that is not necessarily what the user intended. Unlike an edit-time experience, there are some cases where we can't properly suggest all values they may want to type.

So if the user wants to type: `window.some.property` but there is already `window.someother` then when the user types `window.some.`, `someother` will get auto-complete, even though that is not what they wanted. The same thing happens if the user hits Enter to execute what they've already typed.

If you look at how the editor inside the Chrome devtools works, rather than aggressively completing on `.` and Enter, the top suggestion is *only* taken if the user hits Tab or right arrow. However, if the user first arrows up/down in the suggest list, then hitting Enter will complete it (though . still doesn't complete it.)

So in summary, we would like an editor option or hooks into suggest behavior that would allow us to implement the following behavior:

1.  When typing `Enter` and other trigger characters like `.` do not complete the active suggestion.
2.  Right arrow and Tab always complete the active suggestion.
3.  After interacting with the SuggestWidget (via keyboard), `Enter` completes the active suggestion.

If there's some way we can do this today, I'm happy to work around it, but I don't think we have hooks to know if the user has interacted with the SuggestWidget or not.

**Comment by jrieken** [comment]
There are two settings that might help you with this 

```
"editor.acceptSuggestionOnCommitCharacter": true|false,
"editor.acceptSuggestionOnEnter": "on" | "smart" | "off",
```

**Comment by xirzec** [comment]
What does "smart" do? The docs don't enlighten me:

https://microsoft.github.io/monaco-editor/api/interfaces/monaco.editor.ieditoroptions.html#acceptsuggestiononenter

**Comment by jrieken** [comment]
> Controls if suggestions should be accepted on 'Enter' - in addition to 'Tab'. Helps to avoid ambiguity between inserting new lines or accepting suggestions. The value 'smart' means only accept a suggestion with Enter when it makes a textual change

**Comment by xirzec** [comment]
I'm sorry I don't understand what "when it makes a textual change" means.

Specifically we want the behavior that enter will never complete unless the user has moved focus into the SuggestWidget to pick a particular suggestion. 

If the user types "a" and the SuggestWidget's top result is "aVeryLongIdentifier" and the user presses "enter", what happens? Does it leave "a" or does it complete it to "aVeryLongIdentifier"?

**Comment by jrieken** [comment]
> I'm sorry I don't understand what "when it makes a textual change" means.

`console.log|` and `log` is being suggested. When you accept the `log` suggestion no textual change is being made because everything is typed out already. Saying `smart` means a new line is being inserted in those cases. 

> Specifically we want the behavior that enter will never complete unless the user has moved focus into the SuggestWidget to pick a particular suggestion.

I have understood. We don't support that currently. The combination of `editor.acceptSuggestionOnCommitCharacter: false` and `editor.acceptSuggestionOnEnter: off` could help you until we do support that.

**Comment by xirzec** [comment]
Ah I understand now. Yeah the lack of inserting a newline when you had already typed out a suggestion was annoying and I'm glad you have "smart" now for the editor case.

I think our problem with doing what you suggest is it was too confusing for users who arrow down and hit enter expecting the suggestion to be committed but instead get the suggestwidget closed with no completion.
